packages/slate/src/changes/by-key.js

```diff
@@ -422,8 +422,23 @@ Changes.replaceNodeByKey = (change, key, newNode, options = {}) => {
   const node = document.getNode(key)
   const parent = document.getParent(key)
   const index = parent.nodes.indexOf(node)
+  const { selection } = change.value
   change.removeNodeByKey(key, { normalize: false })
   change.insertNodeByKey(parent.key, index, newNode, options)
+  if (selection !== change.value.selection) {
+    const { anchorKey, anchorOffset } = selection
+    const anchorText = newNode.getDescendant(anchorKey)
+    if (anchorText) {
+      const nextAnchorOffset = Math.min(anchorText.text.length, anchorOffset)
+      change.moveAnchorTo(anchorKey, nextAnchorOffset)
+    }
+    const { focusKey, focusOffset } = selection
+    const focusText = newNode.getDescendant(focusKey)
+    if (focusText) {
+      const nextFocusOffset = Math.min(focusText.text.length, focusOffset)
+      change.moveFocusTo(focusKey, nextFocusOffset)
+    }
+  }
```

Comment:
I think there's another case we're actually not handling here too which is if the node is replaced and the key is different, then it won't be selected at all? But ideally it would be selected in the same spot. Would you be down to solve that as well?

Reply:
Do you mean we do not handle the case `focusText.text.length < focusOffset`, and assuming that the text, with the same key,, is unchanged?

My opinion is that, we should be defensive about the case `focusText.text.length < focusOffset` that the text is potentially changed.  There are two options:
1. Do not restore when `focusText.text.length < focusOffset`
2. Restore to the text end

The second solution has one small, even tiny, advantage.  If the {edge}Key of selection is in 
`<emoji isVoid>'SPC'<focus/></emoji>` Then the focusText is reasonable to convert from 'SPC' to ''.  Then the second solution is still good to restore the selection.

Reply:
I agree with you there, I like your solution of doing the `Math.min`.

I think there's another case for this:

```js
<block key="a">
  fo<cursor/>o
</block>

<block key="b">
  ba<cursor/>r
</block>
```

Where the `replaceNodeByKey` is called with a new node that doesn't even share the same key as the previous node. It would be nice if the selection-porting logic preserved that one as well for consistency I think.

Reply:
I think the `bar` does not have the same Text Key with the `foo`; the anchorKey and focusKey are textKeys gotten from the selection before `removeNodeByKey` happens.  Therefore, I think it seems it does not matter at this case?

Reply:
If the text key is not shared, then I think: 
```
+    const anchorText = newNode.getDescendant(anchorKey)
+    if (anchorText) {
```
will prevent the selection restore.  And I think that is perhaps a reasonable logic.

Reply:
@zhujinxuan I think if you're calling `replaceNodeByKey` you're most likely going to want the selection to be preserved in the same way. I believe it will be a lot easier for people to choose to collapse/move than to have to implement preserving it themselves.

Reply:
@ianstormtaylor Do you mean that
```
<block key="a">
<text key="text1">Some<cursor/> </text>
</block>
```
should be converted to 
```
<block key="b">
<text key="text2">Some<cursor/></text>
</block>
```
even the two text keys "text1" and "text2" are different?

I can do it with an optional choice (default turn on preserving by themselves, but allow to disable the preserving if different key )